File: A01 Array/Leetcode 121 Best TimeToBuyAndSellStock/BestTime.java
/*


*/ 
//18ms

class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length < 2) return 0; 
        
        int maxProfit = 0;
        int minPrice = Integer.MAX_VALUE;
        
        for (int price:prices){
            minPrice = Math.min(minPrice, price);
            maxProfit = Math.max(maxProfit, price - minPrice);
        }
        
        return maxProfit;
        
    }
}

////1 ms
class Solution {
    public int maxProfit(int[] prices) {
        // corner case
        if (prices == null || prices.length == 0) return 0;
        
        int minPrice = Integer.MAX_VALUE;
        int maxProfit = 0;
        
        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price;
            }
            else {
                maxProfit = Math.max(maxProfit, price - minPrice);
            }
        }
        
        return maxProfit;
    }
}

File: A01 Array/Leetcode 121 Best TimeToBuyAndSellStock/BestTime.py
def maxProfit(self, prices: List[int]) -> int:
    if len(prices) < 2:
        return 0

    max_profit = 0
    min_buy = prices[0]

    for p in prices[1:]:		
        max_profit = max(max_profit, p - min_buy)
        min_buy = min(min_buy, p)

    return max_profit